appserver/security/core-ee/src/main/java/com/sun/enterprise/security/auth/realm/pam/PamRealm.java

```diff
@@ -77,13 +78,42 @@ public String getAuthType() {
         return AUTH_TYPE;
     }
 
+    public String[] authenticate(String username, String password) {
+        UnixUser user = null;
+        try {
+            user = new PAM(getPamService()).authenticate(username, password);
+        } catch (PAMException e) {
+            _logger.log(SEVERE, "pam_exception_authenticate", e);
+        }
+
+        if (user == null) { // JAAS behavior
+            return null;
+        }
+
+        _logger.log(FINE, () -> "PAM login succeeded for: " + username);
+
+        // Get the groups from the libpam4j UnixUser class that has been returned after a successful authentication.
+
+        String[] groups = null;
+        Set<String> groupSet = user.getGroups();
+
+        if (groupSet != null) {
+            groups = new String[groupSet.size()];
+            user.getGroups().toArray(groups);
```

Comment:
Why not use `groupSet` instead of method call?

Reply:
This was simply the existing code moved from the LoginModule.